base/dict.jl

```diff
@@ -684,7 +684,21 @@ length(t::Dict) = t.count
     (v isa KeySet ? v.dict.keys[i] : v.dict.vals[i], i+1)
 end
 
-filter!(f, d::Dict) = filter_in_one_pass!(f, d)
+function filter!(pred, h::Dict{K,V}) where {K,V}
+    ndel = 0
+    @inbounds for i=1:length(h.slots)
+        if h.slots[i] == 0x01 && !pred(Pair{K,V}(h.keys[i], h.vals[i]))
+            isbitstype(K) || isbitsunion(K) || ccall(:jl_arrayunset, Cvoid, (Any, UInt), h.keys, i-1)
+            isbitstype(V) || isbitsunion(V) || ccall(:jl_arrayunset, Cvoid, (Any, UInt), h.vals, i-1)
+            h.slots[i] = 0x02
+            ndel += 1
```

Comment:
I think I'd prefer calling `_delete!` here to reduce the amount of repeated code.